File: src/main/groovy/net/zomis/brainf/model/ast/tree/ChangeValueSyntax.java
package net.zomis.brainf.model.ast.tree;

import net.zomis.brainf.model.BrainfuckRunner;

public class ChangeValueSyntax extends Syntax implements SteppableSyntax {

    private final int value;

    public ChangeValueSyntax(int value) {
        this.value = value;
    }

    @Override
    public int getValue() {
        return value;
    }

    @Override
    public void perform(BrainfuckRunner runner) {
        performTimes(runner, Math.abs(value));
    }

    @Override
    public String toString() {
        return "ChangeValueSyntax{" +
                "value=" + value +
                '}';
    }

    @Override
    public void performTimes(BrainfuckRunner runner, int steps) {
        steps = value >= 0 ? steps : -steps;
        runner.getMemory().changeMemory(steps);
    }

}

File: brainduck-kotlin/src/commonMain/kotlin/net/zomis/brainf/model/ast/tree/ChangeValueSyntax.kt
package net.zomis.brainf.model.ast.tree;

import net.zomis.brainf.model.BrainfuckRunner
import kotlin.math.absoluteValue

class ChangeValueSyntax(val value: Int) : Syntax(), SteppableSyntax {

    override fun getValue(): Int = value

    override fun perform(runner: BrainfuckRunner) {
        performTimes(runner, value.absoluteValue)
    }

    override fun toString(): String {
        return "ChangeValueSyntax{" +
                "value=" + value +
                '}';
    }

    override fun performTimes(runner: BrainfuckRunner, steps: Int) {
        val steps2 = if (value >= 0) steps else -steps
        runner.memory.changeMemory(steps2)
    }

}
